src/Forms/DiffTransformation.php

```diff
@@ -82,34 +26,23 @@ public function transform(FormField $field)
             return clone $field;
         }
 
-        $data = $this->getComparisonData();
         $name = $field->getName();
 
         // Do not compare generated security data
-        if (!isset($data[$name])
-            && ($form = $field->getForm())
+        if (($form = $field->getForm())
             && ($securityToken = $form->getSecurityToken())
             && ($securityTokenName = $securityToken->getName())
             && $securityTokenName === $name
         ) {
             return LiteralField::create($name, '');
         }
 
-        // Do not "compare apples with oranges"
-#        if (!isset($data[$name])) {
-#            throw new LogicException("No comparison information set for field \"$name\"");
-#        }
-
         try {
             // First check if a field implements performDiffTransformation()
             parent::transform($field);
         } catch (BadMethodCallException $e) {
-            $diffField = HTMLReadonlyField::create(
-                $name,
-                $field->Title(),
-                Diff::compareHTML($field->Value(), isset($this->data[$name]) ? $this->data[$name] : '')
-            );
-
+            $diffField = DiffField::create($name, $field->Title());
```

Comment:
I only just noticed this. We're actually _expecting_ exceptions to be thrown here in order for this class to run its happy path logic? That's wrong... oh well, it is what it is

Reply:
Yeah, the nature of `FormTransformation` is that they actually call `$field->transform($transformation)` rather than having _anything_ to do with `Form`.

In turn `$field->transform($transformation)` calls `$transformation->transform($this)` if and only if there is not a function intended for use with higher precedence (`$field->performDiffTransformation()`)
...so the transformation (class) has nothing at all to do with the `FormField`, either.

There are no hook points, so we rely on the field being trasnformed not knowing what to do with a DiffTransformation first, in order to not duplicate the logic between `FormField` and this `DiffTransformation`.

The whole thing is a bit messy, but indeed, it is what it is.